vault/data_source_ldap_dynamic_role_credentials.go

```diff
@@ -0,0 +1,167 @@
+	return &schema.Resource{
+		ReadContext: ReadContextWrapper(readLDAPDynamicCreds),
+		Schema: map[string]*schema.Schema{
+			consts.FieldMount: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "LDAP Secret Backend to read credentials from.",
+			},
+			consts.FieldRoleName: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the role.",
+			},
+			consts.FieldLeaseID: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Lease identifier assigned by Vault.",
+			},
+			consts.FieldLeaseDuration: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Lease duration in seconds.",
+			},
+			consts.FieldLeaseRenewable: {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "True if the duration of this lease can be extended through renewal.",
+			},
+			consts.FieldDistinguishedNames: {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "List of the distinguished names (DN) created.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+			consts.FieldPassword: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Password for the dynamic role.",
+				Sensitive:   true,
+			},
+			consts.FieldUsername: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Name of the dynamic role.",
+			},
+		},
+	}
+}
+
+func readLDAPDynamicCreds(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := provider.GetClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	mount := d.Get(consts.FieldMount).(string)
+	role := d.Get(consts.FieldRoleName).(string)
+	fullPath := fmt.Sprintf("%s/creds/%s", mount, role)
+
+	secret, err := client.Logical().ReadWithContext(ctx, fullPath)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error reading from Vault: %s", err))
+	}
+	log.Printf("[DEBUG] Read %q from Vault", fullPath)
+	if secret == nil {
+		return diag.FromErr(fmt.Errorf("no role found at %q", fullPath))
+	}
+
+	response, err := parseLDAPDynamicCredSecret(secret)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(response.username)
+	if err := d.Set(consts.FieldLeaseID, response.leaseID); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set(consts.FieldLeaseDuration, response.leaseDuration); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set(consts.FieldLeaseRenewable, response.leaseRenewable); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set(consts.FieldDistinguishedNames, response.distinguishedNames); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set(consts.FieldPassword, response.password); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set(consts.FieldUsername, response.username); err != nil {
+		return diag.FromErr(err)
+	}
```

Comment:
Can we use a loop here? Something similar that was done [here](https://github.com/hashicorp/terraform-provider-vault/blob/283770a5a1be807848999e02aab7635ae47fdeca/vault/resource_ldap_secret_backend.go#L207-L228)

Reply:
although technically possible by converting the response into a `map[string]any{}`, i'm a bit hesitant to do it here because we'd be giving up type safety after parsing & validating the secret.

if others feel like it's worth the tradeoff, i'm more than happy to change it! i know the duplication isn't pretty.

Reply:
Makes sense. I think safety is more important